File: frontend/src/service/complaint.ts
import { postRequest, getRequest } from "./api";

// Evidence/Proof file type
export interface ProofFile {
  file_name: string;
  file_url?: string;
  file_type?: string;
}

// Complaint assignment type
export interface Assignment {
  group: string;
  assigned_by: string;
  worker_id: string;
  status: string;
  assigned_at: string;
  sla_deadline: string;
  retries: number;
  proof_files: ProofFile[];
  verified_by?: string;
  submitted_at?: string;
  verified_at?: string;
  escalate_reason?: string;
  final_note?: string;
}

// Complaint type including full details
export interface ComplaintPayload {
  id: string;
  group?: string;
  full_name: string;
  mobile_number: string;
  email?: string;
  complaint_category: string;
  complaint_subject: string;
  detailed_description: string;
  location: { lat: number; lng: number };
  complete_address: string;
  evidence_files: ProofFile[];
  status: string;
  user_id: string;
  created_at: string;
  assignments: Assignment[];
  current_assignment?: Assignment;
}


export interface CreateComplaintPayload {
  full_name: string;
  mobile_number: string;
  email?: string;
  complaint_category: string;
  complaint_subject: string;
  detailed_description: string;
  complete_address: string;
  location: { lat: number; lng: number };
  evidence_files: ProofFile[];
}

// Assign complaint payload
export interface AssignComplaintPayload {
  group?: string;
  worker_user_id?: string;
  remarks?: string;
  sla_minutes?: number;
}

// Assign complaint response
export interface AssignComplaintResponse {
  message: string;
  complaint: ComplaintPayload;
  available_workers?: Worker[];
}

// Worker types
export interface Worker {
  id: string;
  full_name: string;
  role: string;
  available: boolean;
  last_assigned_at: string;
  created_at: string;
  max_tasks?: number;
  active_tasks?: number;
  can_take_more?: boolean;
}

export interface CreateWorkerPayload {
  full_name: string;
  role: string;
}

// API requests
export const createComplaint = (payload: CreateComplaintPayload) =>
  postRequest("/complaint/", payload);

export const getComplaints = () =>
  getRequest<ComplaintPayload[]>("/complaint/");

// Assign complaint response
export const assignComplaint = (
  complaint_id: string,
  payload: AssignComplaintPayload
) => postRequest<AssignComplaintResponse>(`/complaint/${complaint_id}/assign`, payload);

export const getWorkers = () => getRequest<Worker[]>("/worker/");

export const createWorker = (payload: CreateWorkerPayload) =>
  postRequest<Worker>("/worker/", payload);

File: api/db/models/complaint.py
from enum import Enum
from beanie import Document
from pydantic import BaseModel
from typing import Optional, List, Any
from datetime import datetime

# ------------------ Status Enums ------------------ #
class ComplaintStatus(str, Enum):
    PENDING = "pending"
    VALIDATED = "validated"
    ASSIGNED = "assigned"
    FORWARDED_TO_JE = "forwarded_to_je"
    SUBMITTED_BY_CM = "submitted_by_cm"
    SUBMITTED_BY_JE = "submitted_by_je"
    VERIFIED_BY_JE = "verified_by_je"
    SUBMITTED_BY_CONTRACTOR = "submitted_by_contractor"
    ASSIGNED_TO_CONTRACTOR = "assigned_to_contractor"
    ESCALATED = "escalated"
    CLOSED = "closed"
    REJECTED = "rejected"

class AssignmentStatus(str, Enum):
    ASSIGNED = "assigned"
    SUBMITTED_BY_CM = "submitted_by_cm"
    VERIFIED_BY_JE = "verified_by_je"
    VERIFIED_BY_GM = "verified_by_gm"
    ASSIGNED_TO_CONTRACTOR = "assigned_to_contractor"
    SUBMITTED_BY_CONTRACTOR = "submitted_by_contractor"
    SUBMITTED_BY_JE = "submitted_by_je"
    VERIFIED = "verified"
    ESCALATED = "escalated"
    CANCELLED = "cancelled"

# ------------------ Submodels ------------------ #
class ProofFile(BaseModel):
    file_name: str
    file_url: Optional[str] = None
    file_type: Optional[str] = None
    uploaded_at: datetime = datetime.utcnow()

class Assignment(BaseModel):
    group: Optional[str] = None
    worker_id: Optional[Any] = None
    assigned_by: Optional[Any] = None
    assigned_user_id: Optional[str] = None
    assigned_at: Optional[datetime] = None
    status: AssignmentStatus = AssignmentStatus.ASSIGNED
    sla_deadline: Optional[datetime] = None
    proof_files: Optional[List[ProofFile]] = []
    submitted_at: Optional[datetime] = None
    verified_by: Optional[Any] = None
    verified_at: Optional[datetime] = None
    escalate_reason: Optional[str] = None
    retries: int = 0
    final_note: Optional[str] = None

# ------------------ Worker Model ------------------ #
class Worker(Document):
    full_name: str
    role: str  # plumber, fitter, contractor
    available: bool = True
    last_assigned_at: Optional[datetime] = None
    created_at: datetime = datetime.utcnow()
    max_tasks: Optional[int] = 3
    group: str = "Worker"
    created_by_je: Optional[str] = None

    class Settings:
        name = "workers"

# ------------------ Complaint Model ------------------ #
class Complaint(Document):
    full_name: str
    mobile_number: str
    email: Optional[str] = None
    complaint_category: str
    complaint_subject: str
    detailed_description: str
    location: Optional[dict] = None
    complete_address: str
    evidence_files: Optional[List[ProofFile]] = []
    status: ComplaintStatus = ComplaintStatus.PENDING
    assignments: Optional[List[Assignment]] = []
    current_assignment_index: Optional[int] = None
    created_at: datetime = datetime.utcnow()
    updated_at: datetime = datetime.utcnow()
    user_id: Optional[str] = None  # stores assigned worker's user_id
    forwarded_to_je: Optional[str] = None
    # ------------------ Contractor fields ------------------ #
    assigned_contractor_id: Optional[str] = None
    assigned_by_je: Optional[str] = None
    verified_by_je: Optional[str] = None
    verification_action: Optional[str] = None
    verification_note: Optional[str] = None

    class Settings:
        name = "complaints"
